pallets/portfolio/src/lib.rs

```diff
@@ -0,0 +1,252 @@
+        /// All non-default portfolio numbers and names of a DID.
+        UserPortfolios(IdentityId, Vec<(PortfolioNumber, PortfolioName)>),
+    }
+}
+
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// The portfolio doesn't exist.
+        PortfolioDoesNotExist,
+        /// Insufficient balance for a transaction.
+        InsufficientPortfolioBalance,
+        /// The source and destination portfolios should be different.
+        DestinationIsSamePortfolio,
+        /// The portfolio couldn't be renamed because the chosen name is already in use.
+        PortfolioNameAlreadyInUse,
+    }
+}
+
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        type Error = Error<T>;
+
+        /// The event logger.
+        fn deposit_event() = default;
+
+        /// Creates a portfolio.
+        #[weight = 200_000]
+        pub fn create_portfolio(origin, name: PortfolioName) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            let name_uniq = <Portfolios>::iter_prefix(&did).all(|n| n.1 != name);
+            ensure!(name_uniq, Error::<T>::PortfolioNameAlreadyInUse);
+            let num = Self::get_next_portfolio_number(&did);
+            <Portfolios>::insert(&did, &num, name.clone());
+            Self::deposit_event(RawEvent::PortfolioCreated(did, num, name));
+            Ok(())
+        }
+
+        /// Deletes a user portfolio and moves all its assets to the default portfolio.
+        #[weight = 1_000_000]
+        pub fn delete_portfolio(origin, num: PortfolioNumber) -> DispatchResult {
+            let sender = ensure_signed(origin)?;
+            let did = Context::current_identity_or::<Identity<T>>(&sender)?;
+            ensure!(Self::portfolios(&did, &num).is_some(), Error::<T>::PortfolioDoesNotExist);
+            let portfolio_id = PortfolioId::user_portfolio(did, num);
+            let def_portfolio_id = PortfolioId::default_portfolio(did);
+            for (ticker, balance) in <PortfolioAssetBalances<T>>::iter_prefix(&portfolio_id) {
+                <PortfolioAssetBalances<T>>::mutate(&def_portfolio_id, ticker, |v| {
+                    *v = v.saturating_add(balance)
+                });
+                Self::deposit_event(RawEvent::MovedBetweenPortfolios(
+                    did,
+                    Some(num),
+                    None,
+                    ticker,
+                    balance,
+                ));
+            }
+            <PortfolioAssetBalances<T>>::remove_prefix(&portfolio_id);
+            <Portfolios>::remove(&did, &num);
+            Self::deposit_event(RawEvent::PortfolioDeleted(did, num));
+            Ok(())
+        }
+
+        /// Moves a token amount from one portfolio of an identity to another portfolio of the same
+        /// identity.
+        #[weight = 500_000]
+        pub fn move_portfolio(
```

Comment:
it could be a multi-operation as well. We can also provide the functionality where users can migrate multiple tickers from one portfolio to another.

Reply:
I think we can have just one "move" function making it accept non-empty lists of ticker-amount pairs. It will have the desired batch functionality.